src/app/services/baw-api/baw-api.service.ts

```diff
@@ -255,11 +255,88 @@ export interface Paging {
   next?: string;
 }
 
+interface Combinations {
+  and?: InnerFilter;
+  or?: InnerFilter;
+  not?: InnerFilter;
+}
+
+interface Comparisons {
+  eq?: string | number;
+  equal?: string | number;
+  notEq?: string | number;
+  notEqual?: string | number;
+  lt?: number;
+  lessThan?: number;
+  notLt?: number;
+  notLessThan?: number;
+  gt?: number;
+  greaterThan?: number;
+  notGt?: number;
+  notGreaterThan?: number;
+  lteq?: number;
+  lessThanOrEqual?: number;
+  notLteq?: number;
+  notLessThanOrEqual?: number;
+  gteq?: number;
+  greaterThanOrEqual?: number;
+  notGteq?: number;
+  notGreaterThanOrEqual?: number;
+}
+
+type Range =
+  | string[]
+  | number[]
+  | {
+      interval?: string;
+      from?: number;
+      to?: number;
+    };
+
+interface Subsets {
+  range?: Range;
+  inRange?: Range;
+  notRange?: Range;
+  notInRange?: Range;
+  in?: string[] | number[];
+  notIn?: string[] | number[];
+  contains?: string;
+  contain?: string;
+  notContains?: string;
+  doesNotContain?: string;
+  startsWith?: string;
+  startWith?: string;
+  notStartsWith?: string;
+  notStartWith?: string;
+  doesNotStartWith?: string;
+  endsWith?: string;
+  endWith?: string;
+  notEndsWith?: string;
+  notEndWith?: string;
+  regex?: RegExp;
+  regexMatch?: RegExp;
+  matches?: RegExp;
+  notRegex?: RegExp;
+  notRegexMatch?: RegExp;
+  doesNotMatch?: RegExp;
+  notMatch?: RegExp;
+}
+
+/**
+ * Api response inner filter
+ */
+export type InnerFilter<T = undefined> =
+  | (Combinations &
+      Comparisons &
+      Subsets &
+      { [P in keyof T]?: Combinations & Comparisons & Subsets })
+  | {};
+
```

Comment:
Wow

Reply:
This is getting close to what i want. Typing isn't perfect yet as a programmer can still add keys that aren't part of the model, and values which do not match the key. Both should be possible to fix.

Reply:
Also if i can get this working properly, it should be able to accept empty filters (`{}`). Currently i have a workaround with the `| {};` bit which im not overly happy with.

Reply:
Fixed with the following change:

```typescript
export type InnerFilter<T = {}> = Combinations &
  Comparisons &
  Subsets &
  { [P in keyof T]?: Combinations & Comparisons & Subsets };
```

I still haven't added the ability for the filter to verify the subsets are valid for the key type, however I've decided that'd be over the top.